### DISCUSSION TITLE
Flux2 UI Wireframes Proposal

### DISCUSSION THREAD

**Discussion by jpellizzari** [fluxcd/flux2#1072] [Proposals]
Here is a collection of wireframes that I am proposing we build for the initial Flux2 UI:

![workloads](https://user-images.githubusercontent.com/2802257/110488373-445e7e00-80a3-11eb-984b-23759340ce63.png)
In this "Workloads" view, we are showing users some status information about their Flux-aware workloads, as well as prompting them to onboard their non-Flux-aware workloads.

**Comment by jpellizzari** [comment]
![workload detail timeline clicked](https://user-images.githubusercontent.com/2802257/110491109-cd76b480-80a5-11eb-9b70-4fea1270581f.png)
Here is the workload detail page, with a timeline visualization

**Comment by jpellizzari** [comment]
![workload image automation suggestion](https://user-images.githubusercontent.com/2802257/110491224-ea12ec80-80a5-11eb-9bb8-7a43f9a34c48.png)
Here is an example of further onboarding to be done via the command-line, in this case setting up image automation

**Comment by jpellizzari** [comment]
![unconfigured deployment](https://user-images.githubusercontent.com/2802257/110491311-044cca80-80a6-11eb-9598-2bd3d0af231c.png)
A proposed onboarding flow for non-flux-aware workloads. The color of the numbers indicates status.

**Reply by cdavisafc** [reply]
Is the "context" the Kubectl context?

**Reply by cdavisafc** [reply]
And what is/defines a "workload"?

**Reply by jpellizzari** [reply]
> Is the "context" the Kubectl context?

Yes

>And what is/defines a "workload"?

It is a bit arbitrary, but the K8s objects that classify as "workloads" are listed here: https://kubernetes.io/docs/concepts/workloads/

**Reply by squaremo** [reply]
There are objects other than workloads that

 - might be created by a sync
 - have some sense of health, progress or readiness
 - are interesting to end users

... for example, persistent volume claims. In fact almost everything has some controller which acts on it, and therefore some kind of status and readiness -- so I reckon these views would be useful for almost any object type, even if trivial for some (e.g., ConfigMap).

**Reply by palemtnrider** [reply]
Is the workload just the manifests/helm charts coming from the repo?  i.e., if a manifest in the repo is a CR for an operator and the operator generates additional workloads, what is show here?

**Reply by jpellizzari** [reply]
A workload would map to a `Deployment`, `Job`, or any of the other object types that are listed in the article

**Reply by jpellizzari** [reply]
> There are objects other than workloads that
> 
> * might be created by a sync
> * have some sense of health, progress or readiness
> * are interesting to end users
> 
> ... for example, persistent volume claims. In fact almost everything has some controller which acts on it, and therefore some kind of status and readiness -- so I reckon these views would be useful for almost any object type, even if trivial for some (e.g., ConfigMap).

Every Kubernetes UI (that I have seen), makes decisions around which subset of object types it will support. We will not be able to support all object types (think CRDs), so we will need to decide what is in *our* subset.

**Reply by squaremo** [reply]
> We will not be able to support all object types (think CRDs)

The Kubernetes API is reflective -- you can ask it for the types available to work with. On that level you really can support all object types. If the aim is to reassure the user that their Flux sync is working, I think the user interface has to make all efforts to display accurately everything involved in the sync.

**Reply by jpellizzari** [reply]
Designing a UI that fits all object types is the issue. How can we show a meaningful structured HTML UI if we have an object we have not seen before?

**Reply by hiddeco** [reply]
Given YAML is structured, and there is a standardization for health (`kstatus`). I think it would be possible to create some type of fallback for resources that are of an "unknown" kind?

**Reply by jpellizzari** [reply]
It would need to be so generic, that you may as well render the `yaml` in the page (which could still be valuable). This is the approach I have taken on a similar project.

**Reply by squaremo** [reply]
> that you may as well render the `yaml` in the page

There are standard fields that all objects have, which are enough to identify an object and its type. Then, kstatus can make a pretty good guess at an arbitrary object's ready state, so you have at least `{standard fields, ready state, click for full definition}` to work with.

**Reply by jpellizzari** [reply]
The issue there is that by having to be so generic, you lose the ability to structure the data. The solution I have used in That Other Project™ is to have structured views for "popular" object types, and return the raw definition for the others.

**Comment by resouer** [comment]
Is there a topology graph of the resources?

btw, what is the tool you created this draft may I ask? ^^ @jpellizzari

**Reply by jpellizzari** [reply]
> Is there a topology graph of the resources?

I have some ideas, but nothing formalized yet

> btw, what is the tool you created this draft may I ask?

[Balsamiq](https://balsamiq.cloud/). Works great for fast prototyping

**Comment by squaremo** [comment]
To my mind, there are two "day 1" concerns for people using Flux:

 1. did I set it up right? (or: _what_ did I set up?) and,
 2. what is it doing?

The wireframes in [the original GUI proposal](https://github.com/fluxcd/flux2/discussions/503#discussioncomment-143874) are oriented towards those first, and those above are oriented towards the second.

Concentrating on the wireframes here, one thing I notice is that they don't relate to (specific) syncs -- Kustomizations and whatnot -- so they are less about "what is Flux doing" and more "what is happening, and is Flux involved". Those two projections aren't mutually exclusive, and I can think of uses for both, but if I had to pick one to start with I would pick "what is Flux doing".

**Reply by jpellizzari** [reply]
@squaremo Good call outs; I should have included the parts that are already working in the UI:
![Screenshot from 2021-03-17 10-29-56](https://user-images.githubusercontent.com/2802257/111511374-e74e7200-870b-11eb-89cc-86194e4efdf7.png)
![Screenshot from 2021-03-17 10-30-08](https://user-images.githubusercontent.com/2802257/111511385-ea496280-870b-11eb-807d-dd78ca3f4870.png)

There are already basic list -> detail view and navigation for flux2 primitives. I think they go a long way to answering "what is flux doing", or rather "what is the state of flux"

**Reply by squaremo** [reply]
:+1: Joining these two sets of wireframes/screenshots together would be a firm basis for 2.) above I reckon.

I would think sometimes people will arrive at the UI thinking "Is my sync doing what I think" (screenshot right above, leading into view of the objects it synced); and sometimes they will turn up thinking "where did this (problematic) object come from?" (wireframes at top, leading back to originating syncs/commits).

**Reply by jpellizzari** [reply]
Also added an events page recently that acts as a log:
![Screenshot from 2021-03-17 15-01-26](https://user-images.githubusercontent.com/2802257/111544456-be40d800-8731-11eb-8025-cebcb8d6027c.png)

**Reply by stefanprodan** [reply]
@jpellizzari the events have little value if you don't show the involved object Kind/Namespace/Name.

**Reply by jpellizzari** [reply]
> events have little value if you don't show the involved object Kind/Namespace/Name.

Added in a branch already, Merging soon!

**Comment by palemtnrider** [comment]
I don't see it here in the wireframes, maybe it's out of scope but what is the relationship between workloads and sources?  It seems like a source could be responsible for multiple workloads.  Is that correct?  If so, will the sources page list workloads?

**Reply by jpellizzari** [reply]
@palemtnrider I didn't see an obvious way in the Flux2 data model to tie them together. The Kustomization -> Workload (or other object) has an easily-queryable annotation. I don't believe such an annotation exists for Source -> Workload
